src/command_line/command_line_text.ts

```diff
@@ -0,0 +1,58 @@
+import { normalizeInputString } from "../utils/text";
+
+type TextChange = NoTextChange | TextEndChange | OtherTextChange;
+
+interface TextEndChange {
+    action: "added" | "removed";
+    char: string;
+}
+
+interface NoTextChange {
+    action: "none";
+}
+
+interface OtherTextChange {
+    action: "other";
+}
+
+export function calculateInputAfterTextChange(oldText: string, newText: string): string {
+    const change = diffLineText(oldText, newText);
+    if (change.action === "added") {
+        return normalizeInputString(change.char);
+    } else if (change.action === "removed") {
+        return "<BS>";
+    } else if (change.action === "none") {
+        // If no change, type nothing.
+        return "";
+    } else {
+        // Rewrite the line if it's not a simple change
+        return `<C-u>${normalizeInputString(newText)}`;
+    }
+}
+
+export function diffLineText(oldLine: string, newLine: string): TextChange {
```

Comment:
we have a "diff" library, would it help here? https://github.com/vscode-neovim/vscode-neovim/blob/3d33a4ac09574ecb46602a4c8944bf346e4500b4/package.json#L2159

maybe not, it looks like this function is doing a "compare", not a "diff"

Reply:
Hm, interesting! I hadn't considered using that. My initial mental prototype was to use something like minimum edit distance to reconstruct the string, but aside from being a lot of work for such a small thing, I was worried about possible desync between our image of the command line and neovim's.

I think that might be interesting if we were ever to make this check more complicated. For such a simple "diff" check, I don't think we actually need to rely on something with so much power